src/core/playground/interfaces.ts

```diff
@@ -0,0 +1,21 @@
+import type { Serializer } from "../../bcs";
+import type { HexInput } from "../../types";
+import type { AccountAuthenticator } from "./accountAuthenticator";
+
+export interface Serializable {
+  serialize(serializer: Serializer): void;
+}
+
+export interface PublicKey<TSignature> extends Serializable {
+  verifySignature(message: HexInput, signature: TSignature): boolean;
+}
+
+export interface PrivateKey<TSignature, TPublicKey extends PublicKey<TSignature>> {
+  sign(message: HexInput): TSignature;
+
+  publicKey(): TPublicKey;
+}
+
+export interface BaseSigner<TAccountAuthenticator extends AccountAuthenticator = AccountAuthenticator> {
```

Comment:
is this to support legacy ed/multied and single signer?

Reply:
This would be a common interface between a `Signer` and a `LegacyEd25519Signer`.
Having this defined is possibly not required. It can be useful to allow passing any signer to a method (e.g. `signTransaction`), but we'll see if/when integrating with the rest of the code if this should stay or go

Reply:
BaseSigner could be confusing in the case of a multisigner, but what I would add is:

1. sign
2. verify
3. public key

Any signer should be able to sign and verify any blob, and show proof that they can verify it (especially since they need the public key anyways)

Reply:
I don't think it makes sense to have a multi-key signer.
The flow is fundamentally different and it would only makes sense if a dapp owns all the secret keys.

For multi-key flows (whether it's multi-sig or multi-agent) we have individual single-key signers signing a common message, and then we combine the signatures.
We should aim to facilitate that (hence the `authenticators` overload in `LegacyMultiEd25519AccountAuthenticator`)

Reply:
All signers should be able to verify and get the public keys associated, single or multi (otherwise how do you verify).

Single signers should be able to sign (you have the key)

Sign shouldn't be applicable if there was a type called MultiKeySigner, since it doesn't have all the keys.

Reply:
> otherwise how do you verify

@xbtmatt and I were talking about possibly always using authenticators to pass signatures around.

An authenticator is typically a signature + publicKey and has all it needs to verify a message already.
I added a `verify(message): boolean` method to authenticators.

https://github.com/aptos-labs/aptos-ts-sdk/blob/e5e717e90002d0a0cfed4e305bbe854ec68ae107/src/core/playground/accountAuthenticator/singleKey.ts#L16-L18

So the flows (for single key) would be:
```typescript

// Signer side
function handleSignatureRequest(message: Uint8Array) {
  const authenticator = signer.sign(message);
  return authenticator;
}

// Verifier side (NO access to signing capabilities)
const authenticator = requestSignature(message); 

// either trust the response (the fullnode would eventually catch the mismatch)
const isVerified = authenticator.verify(message);

// or manually verify (needs to know the public key locally)
const publicKey = getPublicKey();
if (!isSameSignatureScheme(publicKey, authenticator) {
  throw new Error('Signature scheme mismatch');
}
const isVerified = publicKey.verifySignature(authenticator.signature);
```

We can move away from the idea of always using authenticators, and instead always returning signatures.
In that case I would suggest having a different class / interface we could call account and which has address and public key.

```typescript
interface Account<TSignature, TPublicKey<TSignature>> {
  readonly address: string;
  readonly publicKey: TPublicKey;
  
  verifySignature(signature: TSignature): boolean;
}

interface Signer<TSignatrure, TPublicKey<TSignature>> extends Account<TSignature, TPublicKey> {
  sign(message: Uint8Array): TSignature;
}
```

This would "encourage" dapp developers to keep track of the public keys and use them to verify the signature.

Reply:
https://github.com/aptos-labs/aptos-ts-sdk/pull/147#issuecomment-1789871454

Reply:
there is a benefit to return authenticator because it also holds the publicKey